### PR TITLE
[WIP] Preserve singleton type when inferring override

### PR DESCRIPTION
**Pull request by som-snytt**
Fixes scala/bug#12621

### CODE REVIEW

**Comment by som-snytt** [comment]
I expected something like this simple retronymism from ten years ago ("Better return type inheritance for dep. method types"):
```
          // SI-7668 Substitute parameters from the parent method with those of the overriding method.
          overriddenTp = overriddenTp.substSym(overridden.paramss.flatten, vparamss.flatten.map(_.symbol))
```
which currently reads
```
            // scala/bug#7668 Substitute parameters from the parent method with those of the overriding method.
            val overriddenResTp = applyFully(overriddenTp, vparamSymss).substSym(overriddenTparams, tparamSkolems)
```
where `applyFully` is the locus of intervention.

(First swing was to unpack the existential in `assignTypeToTree`. The current swing avoids using `resultType` which does the `ExistentialExtrapolation`. I did not formulate an escape hatch in `existentialAbstraction` that didn't break the more usual dependent case that returns `x.T` instead of `x.type`.)

Quick test, the (correct) escape hatch at `existentialAbstraction` means inferring `sb.type` but we need `xb.type` for typechecking RHS.
```
t12621.scala:13: error: type mismatch;
 found   : xb.type (with underlying type StringBuilder)
 required: sb.type
  override def f(xb: StringBuilder) = xb    // inferred StringBuilder
```

Adding a neg test that shows the correct required type.

I notice the other PR doesn't fail that compilation.

File: src/reflect/scala/reflect/internal/tpe/TypeMaps.scala [hunk 1]
```diff
@@ -27,14 +27,13 @@ private[internal] trait TypeMaps {
   import definitions._
 
   /** Normalize any type aliases within this type (@see Type#normalize).
-    *  Note that this depends very much on the call to "normalize", not "dealias",
-    *  so it is no longer carries the too-stealthy name "deAlias".
-    */
+   *  Note that this depends very much on the call to "normalize", not "dealias",
+   *  so it no longer carries the too-stealthy name "deAlias".
```
Comment (som-snytt) [inline src/reflect/scala/reflect/internal/tpe/TypeMaps.scala]: historical note

File: src/compiler/scala/tools/nsc/typechecker/Namers.scala [hunk 2]
```diff
@@ -1413,13 +1413,25 @@ trait Namers extends MethodSynthesis {
...
-            }
-            else applyFully(tp.resultType(paramss.head.map(_.tpe)), paramss.tail)
+           @tailrec @inline def applyFully(tp: Type, paramss: List[List[Symbol]]): Type =
+             paramss match {
+               case ps :: rest =>
+                 val resTp =
+                   tp match {
+                     case MethodType(qs, SingleType(NoPrefix, sym)) if qs.contains(sym) =>
+                       val i = qs.indexWhere(_ == sym)
+                       singleType(NoPrefix, ps(i))
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/Namers.scala:1423]: This is the relevant new case. Is it sufficient? It seems to handle `f(p): p.type` but not all dependencies.

**Comment by som-snytt** [comment]
This does seem to be an inconsistent test:
```
% diff /home/jenkins/workspace/scala-2.13.x-validate-main/test/files/presentation/doc.check /home/jenkins/workspace/scala-2.13.x-validate-main/test/files/presentation/doc-presentation.log
@@ -1,2 +1,3 @@
 warning: 1 deprecation (since 2.13.0); re-run with -deprecation for details
 reload: Base.scala, Class.scala, Derived.scala
+Unexpected foo method comment:None
```

**Comment by lrytz** [comment]
> +Unexpected foo method comment:None

yes, seen that before

**Pull request closed without merging**